chezmoi crashes with --keep-going

## Describe the bug

I tried to use your suggestion to see what files are going to change using `--keep-going`. I got 2 different crashes depending on whether pager (delta in this case) was installed. Without `-k` it doesn't crash. This can be related to https://github.com/twpayne/chezmoi/issues/3995 as well.

## To reproduce

I don't have small repro, I have source tree with several templates unable to render. Hope these stacktraces give some clues.

`delta` isn't installed. Basically `w.openFunc` is `nil` here.
```console
$ chezmoi apply --dry-run -k -v
chezmoi: .chezmoiscripts/unix/71-install-uv-tools.sh: exec: "delta": executable file not found in $PATH
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xef48a8]

goroutine 1 [running]:
github.com/twpayne/chezmoi/v2/internal/cmd.(*lazyWriter).Write(0x400099a828, {0x40009900a0, 0x44, 0x50})
    /home/runner/work/chezmoi/chezmoi/internal/cmd/lazywriter.go:26 +0x38
github.com/go-git/go-git/v5/plumbing/format/diff.(*UnifiedEncoder).Encode(0x4000339100, {0x1776d80, 0x400026b5c0})
    /home/runner/go/pkg/mod/github.com/go-git/go-git/v5@v5.12.0/plumbing/format/diff/unified_encoder.go:96 +0x29c
github.com/twpayne/chezmoi/v2/internal/chezmoi.(*GitDiffSystem).encodeDiff(0x4000339140, {0x400013a1c8, 0x16}, {0x0, 0x0, 0x0}, 0x800001fd)
    /home/runner/work/chezmoi/chezmoi/internal/chezmoi/gitdiffsystem.go:316 +0x354
github.com/twpayne/chezmoi/v2/internal/chezmoi.(*GitDiffSystem).Chmod(0x4000339140, {0x400013a1c8, 0x16}, 0x1fd)
    /home/runner/work/chezmoi/chezmoi/internal/chezmoi/gitdiffsystem.go:75 +0x10c
github.com/twpayne/chezmoi/v2/internal/chezmoi.(*TargetStateDir).Apply(0x400091a680?, {0x17948c0?, 0x4000339140?}, {0x16?, 0x0?}, {0x1780fa0?, 0x400090db90?})
    /home/runner/work/chezmoi/chezmoi/internal/chezmoi/targetstateentry.go:165 +0x8c
github.com/twpayne/chezmoi/v2/internal/chezmoi.(*SourceState).Apply(0x400065da00, {0x17948c0, 0x4000339140}, {0x17948c0, 0x4000339140}, {0x178b0b8, 0x40000804e0}, {0x40000400c5, 0xe}, {{0x400091a680?, ...}}, ...)
    /home/runner/work/chezmoi/chezmoi/internal/chezmoi/sourcestate.go:771 +0x5c8
github.com/twpayne/chezmoi/v2/internal/cmd.(*Config).applyArgs(0x4000275008, {0x1780e18, 0x287f420}, {0x17948c0, 0x4000339140}, {0x40000400c5, 0xe}, {0x4000347860, 0x0, 0x3}, ...)
    /home/runner/work/chezmoi/chezmoi/internal/cmd/config.go:663 +0x528
github.com/twpayne/chezmoi/v2/internal/cmd.(*Config).runApplyCmd(0x4000275008, 0xcd6a48c800000000?, {0x4000347860, 0x0, 0x3})
    /home/runner/work/chezmoi/chezmoi/internal/cmd/applycmd.go:39 +0xec
github.com/spf13/cobra.(*Command).execute(0x40001a1808, {0x4000347800, 0x3, 0x3})
    /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:985 +0x834
github.com/spf13/cobra.(*Command).ExecuteC(0x400003d508)
    /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1117 +0x344
github.com/spf13/cobra.(*Command).Execute(...)
    /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1041
github.com/twpayne/chezmoi/v2/internal/cmd.(*Config).execute(0x400081fdd0?, {0x400003a0b0, 0x4, 0x4})
    /home/runner/work/chezmoi/chezmoi/internal/cmd/config.go:1244 +0x70
github.com/twpayne/chezmoi/v2/internal/cmd.runMain({{0x1768c98, 0x6}, {0x177d4a0, 0x28}, {0x176ec00, 0x14}, {0x1769b68, 0xa}}, {0x400003a0b0, 0x4, ...})
    /home/runner/work/chezmoi/chezmoi/internal/cmd/cmd.go:271 +0x178
github.com/twpayne/chezmoi/v2/internal/cmd.Main({{0x1768c98, 0x6}, {0x177d4a0, 0x28}, {0x176ec00, 0x14}, {0x1769b68, 0xa}}, {0x400003a0b0?, 0x4a874?, ...})
    /home/runner/work/chezmoi/chezmoi/internal/cmd/cmd.go:107 +0x54
main.main()
    /home/runner/work/chezmoi/chezmoi/main.go:30 +0xe8
 ```

`delta` is installed:

```console
$ chezmoi apply --dry-run -v -k
fatal error: all goroutines are asleep - deadlock!

goroutine 1 [select]:
io.(*pipe).write(0x40001a9da0, {0x4000ae0000, 0x19684, 0x1a000})
    /opt/hostedtoolcache/go/1.23.2/x64/src/io/pipe.go:86 +0x1ac
io.(*PipeWriter).Write(0x40001df8d8?, {0x4000ae0000?, 0x1a000?, 0x4000ae0000?})
    /opt/hostedtoolcache/go/1.23.2/x64/src/io/pipe.go:161 +0x24
github.com/twpayne/chezmoi/v2/internal/cmd.(*lazyWriter).Write(0x4000576410?, {0x4000ae0000?, 0x19684?, 0x40004be2c0?})
    /home/runner/work/chezmoi/chezmoi/internal/cmd/lazywriter.go:33 +0xb0
github.com/go-git/go-git/v5/plumbing/format/diff.(*UnifiedEncoder).Encode(0x40003ad080, {0x1776d80, 0x400016e450})
    /home/runner/go/pkg/mod/github.com/go-git/go-git/v5@v5.12.0/plumbing/format/diff/unified_encoder.go:96 +0x29c
github.com/twpayne/chezmoi/v2/internal/chezmoi.(*GitDiffSystem).encodeDiff(0x40003ad0c0, {0x4000758ac0, 0x34}, {0x400070c000, 0x15989, 0x1598a}, 0x1b4)
    /home/runner/work/chezmoi/chezmoi/internal/chezmoi/gitdiffsystem.go:316 +0x354
github.com/twpayne/chezmoi/v2/internal/chezmoi.(*GitDiffSystem).WriteFile(0x40003ad0c0, {0x4000758ac0, 0x34}, {0x400070c000, 0x15989, 0x1598a}, 0x1b4)
    /home/runner/work/chezmoi/chezmoi/internal/chezmoi/gitdiffsystem.go:243 +0x5c
github.com/twpayne/chezmoi/v2/internal/chezmoi.(*TargetStateFile).Apply(0x400016e2a0, {0x17948c0, 0x40003ad0c0}, {0x34?, 0x0?}, {0x1780f30, 0x40006b6100})
    /home/runner/work/chezmoi/chezmoi/internal/chezmoi/targetstateentry.go:234 +0x284
github.com/twpayne/chezmoi/v2/internal/chezmoi.(*SourceState).Apply(0x4000495a00, {0x17948c0, 0x40003ad0c0}, {0x17948c0, 0x40003ad0c0}, {0x178b0b8, 0x4000199200}, {0x40000400c5, 0xe}, {{0x40004fbb00?, ...}}, ...)
    /home/runner/work/chezmoi/chezmoi/internal/chezmoi/sourcestate.go:771 +0x5c8
github.com/twpayne/chezmoi/v2/internal/cmd.(*Config).applyArgs(0x40000af008, {0x1780e18, 0x287f420}, {0x17948c0, 0x40003ad0c0}, {0x40000400c5, 0xe}, {0x40000b10e0, 0x0, 0x3}, ...)
    /home/runner/work/chezmoi/chezmoi/internal/cmd/config.go:663 +0x528
github.com/twpayne/chezmoi/v2/internal/cmd.(*Config).runApplyCmd(0x40000af008, 0x8eb0f8a100000000?, {0x40000b10e0, 0x0, 0x3})
    /home/runner/work/chezmoi/chezmoi/internal/cmd/applycmd.go:39 +0xec
github.com/spf13/cobra.(*Command).execute(0x400003d808, {0x40000b1080, 0x3, 0x3})
    /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:985 +0x834
github.com/spf13/cobra.(*Command).ExecuteC(0x40001a5808)
    /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1117 +0x344
github.com/spf13/cobra.(*Command).Execute(...)
    /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1041
github.com/twpayne/chezmoi/v2/internal/cmd.(*Config).execute(0x400069fdd0?, {0x4000180060, 0x4, 0x4})
    /home/runner/work/chezmoi/chezmoi/internal/cmd/config.go:1244 +0x70
github.com/twpayne/chezmoi/v2/internal/cmd.runMain({{0x1768c98, 0x6}, {0x177d4a0, 0x28}, {0x176ec00, 0x14}, {0x1769b68, 0xa}}, {0x4000180060, 0x4, ...})
    /home/runner/work/chezmoi/chezmoi/internal/cmd/cmd.go:271 +0x178
github.com/twpayne/chezmoi/v2/internal/cmd.Main({{0x1768c98, 0x6}, {0x177d4a0, 0x28}, {0x176ec00, 0x14}, {0x1769b68, 0xa}}, {0x4000180060?, 0x4a874?, ...})
    /home/runner/work/chezmoi/chezmoi/internal/cmd/cmd.go:107 +0x54
main.main()
    /home/runner/work/chezmoi/chezmoi/main.go:30 +0xe8
```

## Expected behavior

Chezmoi doesn't crash.

## Output of `chezmoi doctor`

<details>

```console
$ ./myczm doctor
RESULT    CHECK                       MESSAGE
warning   version                     dev, commit b8610d493ab06b832a8a924bbf59310c22d37a83, built at 2024-10-25T17:25:14Z
warning   latest-version              v2.52.4
ok        os-arch                     linux/arm64 (Ubuntu 24.04.1 LTS (Noble Numbat))
ok        uname                       Linux UbuntuServer24 6.8.0-47-generic #47-Ubuntu SMP PREEMPT_DYNAMIC Fri Sep 27 22:03:50 UTC 2024 aarch64 aarch64 aarch64 GNU/Linux
ok        go-version                  go1.22.2 (gc)
ok        executable                  ~/chezmoi/myczm
ok        upgrade-method              replace-executable
ok        config-file                 ~/.config/chezmoi/chezmoi.toml, last modified 2024-10-25T17:44:15Z
ok        source-dir                  ~/.local/share/chezmoi is a git working tree (clean)
ok        suspicious-entries          no suspicious entries
ok        working-tree                ~/.local/share/chezmoi is a git working tree (clean)
ok        dest-dir                    ~ is a directory
ok        umask                       002
ok        cd-command                  found /bin/bash
ok        cd-args                     /bin/bash
info      diff-command                not set
ok        edit-command                found /usr/bin/vi
ok        edit-args                   /usr/bin/vi
ok        git-command                 found /usr/bin/git, version 2.43.0
ok        merge-command               found /usr/bin/nvim
ok        shell-command               found /bin/bash
ok        shell-args                  /bin/bash
info      age-command                 age not found in $PATH
ok        gpg-command                 found /usr/bin/gpg, version 2.4.4
info      pinentry-command            not set
info      1password-command           op not found in $PATH
info      bitwarden-command           bw not found in $PATH
info      bitwarden-secrets-command   bws not found in $PATH
info      dashlane-command            dcli not found in $PATH
info      doppler-command             doppler not found in $PATH
info      gopass-command              gopass not found in $PATH
info      keepassxc-command           keepassxc-cli not found in $PATH
info      keepassxc-db                not set
info      keeper-command              keeper not found in $PATH
info      lastpass-command            lpass not found in $PATH
info      pass-command                pass not found in $PATH
info      passhole-command            ph not found in $PATH
info      rbw-command                 rbw not found in $PATH
info      vault-command               vault not found in $PATH
info      vlt-command                 vlt not found in $PATH
info      secret-command              not set

```

</details>

## Additional context

I verified this also happens on the latest master.

--- a/internal/cmd/lazywriter.go
+++ b/internal/cmd/lazywriter.go
@@ -1,11 +1,16 @@
 package cmd
 
-import "io"
+import (
+	"io"
+	"sync"
+)
 
 // A lazyWriter only opens its destination on first write.
 type lazyWriter struct {
+	mutex       sync.Mutex
 	openFunc    func() (io.WriteCloser, error)
 	writeCloser io.WriteCloser
+	err         error
 }
 
 func newLazyWriter(openFunc func() (io.WriteCloser, error)) *lazyWriter {
@@ -15,6 +20,8 @@ func newLazyWriter(openFunc func() (io.WriteCloser, error)) *lazyWriter {
 }
 
 func (w *lazyWriter) Close() error {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
 	if w.writeCloser == nil {
 		return nil
 	}
@@ -22,13 +29,14 @@ func (w *lazyWriter) Close() error {
 }
 
 func (w *lazyWriter) Write(p []byte) (int, error) {
-	if w.writeCloser == nil {
-		writeCloser, err := w.openFunc()
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+	if w.openFunc != nil {
+		w.writeCloser, w.err = w.openFunc()
 		w.openFunc = nil
-		if err != nil {
-			return 0, err
-		}
-		w.writeCloser = writeCloser
+	}
+	if w.err != nil {
+		return 0, w.err
 	}
 	return w.writeCloser.Write(p)
 }

--- a/internal/cmd/testdata/scripts/issue4027.txtar
+++ b/internal/cmd/testdata/scripts/issue4027.txtar
@@ -1,0 +1,13 @@
+# test that chezmoi diff --keep-going does not panic when the diff pager command does not exist
+! exec chezmoi diff --keep-going
+! stderr panic
+[!windows] stderr '/not/exist: no such file or directory'
+[windows] stderr '"/not/exist": executable file not found in %PATH%'
+
+-- home/user/.config/chezmoi/chezmoi.toml --
+[diff]
+    pager = "/not/exist"
+-- home/user/.local/share/chezmoi/dot_file1 --
+# contents of .file1
+-- home/user/.local/share/chezmoi/dot_file2 --
+# contents of .file2

